javalib/src/main/scala/java/lang/Integer.scala

```diff
@@ -50,127 +48,445 @@ object Integer {
+
+  @inline def numberOfLeadingZeros(i: scala.Int): scala.Int =
+    Intrinsics.`llvm.ctlz.i32`(i, iszeroundef = false)
+
+  @inline def numberOfTrailingZeros(i: scala.Int): scala.Int =
+    Intrinsics.`llvm.cttz.i32`(i, iszeroundef = false)
+
+  @inline def parseInt(s: String): scala.Int =
+    parseInt(s, 10)
+
+  def parseInt(s: String, radix: scala.Int): scala.Int = {
+    if (s == null || radix < Character.MIN_RADIX ||
+        radix > Character.MAX_RADIX) throw new NumberFormatException(s)
+    val length   = s.length()
+    val negative = s.charAt(0) == '-'
+
+    if (length == 0)
+      throw new NumberFormatException(s)
+    else if (negative && length == 1)
+      throw new NumberFormatException(s)
+    else
+      parse(s, 1, radix, negative)
+  }
+
+  private def parse(s: String,
+                    _offset: scala.Int,
+                    radix: scala.Int,
+                    negative: scala.Boolean): scala.Int = {
+    val max    = MIN_VALUE / radix
+    val length = s.length()
+    var result = 0
+    var offset = _offset
+
+    while (offset < length) {
+      val digit = Character.digit(s.charAt(offset), radix)
+      offset += 1
+      if (digit == -1) throw new NumberFormatException(s)
+      if (max > result) throw new NumberFormatException(s)
+
+      val next = result * radix - digit
+      if (next > result) throw new NumberFormatException(s)
+      result = next
+    }
+
+    if (!negative) {
+      result = -result
+      if (result < 0) throw new NumberFormatException(s)
+    }
+
+    result
   }
 
+  @inline
+  def remainderUnsigned(dividend: scala.Int, divisor: scala.Int): scala.Int =
+    remUInt(dividend, divisor)
+
+  @inline def reverse(i: scala.Int): scala.Int =
+    Intrinsics.`llvm.bitreverse.i32`(i)
+
+  @inline def reverseBytes(i: scala.Int): scala.Int =
+    Intrinsics.`llvm.bswap.i32`(i)
+
   @inline def rotateLeft(i: scala.Int, distance: scala.Int): scala.Int =
     (i << distance) | (i >>> -distance)
 
   @inline def rotateRight(i: scala.Int, distance: scala.Int): scala.Int =
     (i >>> distance) | (i << -distance)
 
   @inline def signum(i: scala.Int): scala.Int =
-    if (i == 0) 0 else if (i < 0) -1 else 1
-
-  // Intrinsic
-  def numberOfLeadingZeros(i: scala.Int): scala.Int = {
-    // See Hacker's Delight, Section 5-3
-    var x = i
-    if (x == 0) {
-      32
+    if (i == 0) 0
+    else if (i < 0) -1
+    else 1
+
+  @inline def sum(a: scala.Int, b: scala.Int): scala.Int =
+    a + b
+
+  def toBinaryString(i: scala.Int): String = {
+    val count  = if (i == 0) 1 else 32 - numberOfLeadingZeros(i)
+    val buffer = new Array[Char](count)
+    var k = i
+    do {
+      count -= 1
+      buffer(count) = ((k & 1) + '0').toChar
+      k >>>= 1
+    } while (count > 0)
+
+    new String(buffer)
+  }
+
+  def toHexString(i: scala.Int): String = {
+    val count =
+      if (i == 0) 1
+      else {
+        val ilz = 32 - numberOfLeadingZeros(i)
+        ilz / 4 + ilz % 4
+      }
```

Comment:
Hum, nope, that's not correct. When `i == 2`, `ilz == 2` so your formula gives `count == 2`, but it should be 1.

It should be `((32 - numberOfLeadingZeros(i)) + 3) / 4`. Essentially it's integer division but with rounding *up* instead of rounding *down*.